include/irr/asset/CTexturePacker.h

```diff
@@ -53,20 +55,119 @@ class CSquareAddressAllocator
     }
 };
 
-class CTexturePacker
+class ITexturePacker
 {
-    _IRR_STATIC_INLINE_CONSTEXPR uint32_t TILE_PADDING = 8u;
-
+protected:
     _IRR_STATIC_INLINE_CONSTEXPR uint32_t ADDR_LAYER_SHIFT = 28u;//14bit 2d page address and 4bit for 3rd address dimension (layer)
     _IRR_STATIC_INLINE_CONSTEXPR uint32_t ADDR_MORTON_MASK = (1u<<ADDR_LAYER_SHIFT)-1u;
 
-    using addr_alctr_t = core::GeneralpurposeAddressAllocator<uint32_t>;
+    const uint32_t m_pgSzxy;
+    const uint32_t m_tilesPerDim;
+    CSquareAddressAllocator<uint32_t> m_addrAlctr;
+
+    core::smart_refctd_ptr<ICPUImage> m_pageTable;
+public:
+    struct STexOffset
+    {
+        _IRR_STATIC_INLINE_CONSTEXPR uint32_t X_MASK = ((1u<<(ADDR_LAYER_SHIFT>>1))-1u);
+        _IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_addr = (~0u);
+
+        STexOffset(uint32_t _addr) : addr(_addr) {}
+        STexOffset(uint32_t _x, uint32_t _y, uint32_t _layer)
+        {
+            addr = (_x&X_MASK);
+            addr |= (_y&X_MASK)<<(ADDR_LAYER_SHIFT>>1);
+            addr |= (_layer<<ADDR_LAYER_SHIFT);
+        }
+        uint32_t x() const { return addr&X_MASK; }
+        uint32_t y() const { return (addr>>(ADDR_LAYER_SHIFT>>1))&X_MASK; }
+        uint32_t layer() const { return addr>>ADDR_LAYER_SHIFT; }
+
+        uint32_t addr;
+    };
+
+    ITexturePacker(uint32_t _pgSzxy = 256u, uint32_t _tilesPerDim = 32u, uint32_t _numLayers = 4u) :
+        m_pgSzxy(_pgSzxy),
+        m_tilesPerDim(_tilesPerDim),
+        m_addrAlctr(14u, _tilesPerDim, _numLayers)
+    {
+        ICPUImage::SCreationParams params;
+        params.arrayLayers = _numLayers;
+        params.extent = {_tilesPerDim,_tilesPerDim,1u};
+        params.flags = static_cast<ICPUImage::E_CREATE_FLAGS>(0);
+        params.format = EF_R32_UINT;
+        params.mipLevels = core::findMSB(_tilesPerDim)+1;
```

Comment:
take the max number of mip levels as a parameter (this will also determine the max texture size you can allocate/pack (`MaxTextureSize=m_pgSzxy<<(mipLevels-1u)`)

Reply:
still relevant